src/Data/List/Properties.agda

```diff
@@ -163,9 +169,32 @@ module _ {a} {A : Set a} where
   ++-identityˡ-unique {xs = x ∷ xs} (y ∷ []   ) eq | ()
   ++-identityˡ-unique {xs = x ∷ xs} (y ∷ _ ∷ _) eq | ()
 
-  length-++ : ∀ (xs : List A) {ys} → length (xs ++ ys) ≡ length xs + length ys
-  length-++ []       = refl
-  length-++ (x ∷ xs) = P.cong suc (length-++ xs)
+  ++-cancelˡ : ∀ xs {ys zs : List A} → xs ++ ys ≡ xs ++ zs → ys ≡ zs
+  ++-cancelˡ []       ys≡zs             = ys≡zs
+  ++-cancelˡ (x ∷ xs) x∷xs++ys≡x∷xs++zs = ++-cancelˡ xs (∷-injectiveʳ x∷xs++ys≡x∷xs++zs)
+
+  ++-cancelʳ : ∀ {xs} ys zs → ys ++ xs ≡ zs ++ xs → ys ≡ zs
+  ++-cancelʳ []       []       _             = refl
+  ++-cancelʳ {xs} []           (z ∷ zs) eq =
+    contradiction (P.trans (cong length eq) (length-++ (z ∷ zs))) (m≢1+n+m (length xs))
+  ++-cancelʳ {xs} (y ∷ ys) []       eq =
+    contradiction (P.trans (P.sym (length-++ (y ∷ ys))) (cong length eq)) (m≢1+n+m (length xs) ∘ sym)
+  ++-cancelʳ (y ∷ ys) (z ∷ zs) eq =
+    P.cong₂ _∷_ (∷-injectiveˡ eq) (++-cancelʳ ys zs (∷-injectiveʳ eq))
+
+  ++-cancel : Cancellative _++_
+  ++-cancel = ++-cancelˡ , ++-cancelʳ
+
+  ++-conicalˡ : ∀ (xs ys : List A) → xs ++ ys ≡ [] → xs ≡ []
```

Comment:
Shouldn't the type read `LeftConical [] _++_`? (Same for `RightConical` below)

Reply:
Oh, is this what you mean by new style? I thought you meant mentioning the types
explicitly in the definition of `XConical`.

Reply:
Yes, so the type is written out explicitly so users can work out what it means. At the same time the proof of `++-conical` ensures that both the left and right versions must have the right type.